mlflow/utils/search_utils.py

```diff
@@ -994,3 +923,108 @@ def _apply_reversor(experiment, key, ascending):
+
+    @classmethod
+    def _get_value(cls, identifier_type, key, token):
+        if identifier_type == cls._TAG_IDENTIFIER:
+            if token.ttype in cls.STRING_VALUE_TYPES or isinstance(token, Identifier):
+                return cls._strip_quotes(token.value, expect_quoted_value=True)
+            raise MlflowException(
+                "Expected a quoted string value for "
+                "{identifier_type} (e.g. 'my-value'). Got value "
+                "{value}".format(identifier_type=identifier_type, value=token.value),
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+        elif identifier_type == cls._ATTRIBUTE_IDENTIFIER:
+            if token.ttype in cls.STRING_VALUE_TYPES or isinstance(token, Identifier):
+                return cls._strip_quotes(token.value, expect_quoted_value=True)
+            elif isinstance(token, Parenthesis):
+                cls._check_valid_identifier_list(token)
+                if key != "run_id":
+                    raise MlflowException(
+                        "Only run_id attribute support compare with a list of quoted string "
+                        "values.",
+                        error_code=INVALID_PARAMETER_VALUE,
+                    )
+                run_id_list = cls._parse_list_from_sql_token(token)
+                # Because MySQL IN clause is case-insensitive, but all run_ids only contains lower
+                # case letters, so that we filter out run_ids containing upper case letters here.
+                run_id_list = [run_id for run_id in run_id_list if run_id.islower()]
+                return run_id_list
+            else:
+                raise MlflowException(
+                    "Expected a quoted string value or a list of quoted string values for "
+                    "attributes. Got value {value}".format(value=token.value),
+                    error_code=INVALID_PARAMETER_VALUE,
+                )
+        else:
+            # Expected to be either "param" or "metric".
+            raise MlflowException(
+                "Invalid identifier type. Expected one of "
+                "{}.".format([cls._ATTRIBUTE_IDENTIFIER, cls._TAG_IDENTIFIER]),
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+    @classmethod
+    def _invalid_statement_token_search_model_registry(cls, token):
+        if isinstance(token, (Comparison, Identifier, Parenthesis)):
+            return False
+        elif token.is_whitespace:
+            return False
+        elif token.match(ttype=TokenType.Keyword, values=["AND", "IN"]):
+            return False
+        elif token.match(ttype=TokenType.Operator.Comparison, values=["IN"]):
+            # `IN` is a comparison token in sqlparse >= 0.4.0:
+            # https://github.com/andialbrecht/sqlparse/pull/567
+            if token.value == "IN":
+                # This case it represent the IN filter parsed failed.
+                # e.g. "run_id IN"
+                return True
```

Comment:
Does this happen for other operators? What happens if `filter_string` looks like `tag.a LIKE`?

Reply:
Parsing `tag.a LIKE` works fine (the parser will raise correct exception),
because if filter is `tag.a LIKE`, then it will run into L1030
the L1021 branch is for backwards compatibility on lsqlparse<0.4 package, and the "run_id IN" filter accidently run into this branch, so the issue only happen on "IN" operator.

Reply:
Can we add a test case for this, if one does not exist?

Reply:
added.